File: my-folder/0067-add-binary/solution.c
char* addBinary(char* a, char* b) {
    size_t a_len = strlen(a), b_len = strlen(b);
    size_t result_len = (a_len > b_len ? a_len : b_len) + 2;
    char *result = calloc(result_len, 1);

    uint8_t carry = 0;
    for (size_t i = result_len - 2; i > 0; i--) {
        uint8_t ca = a_len > 0 ? a[--a_len] - '0' : 0;
        uint8_t cb = b_len > 0 ? b[--b_len] - '0' : 0;

        uint8_t sum = ca + cb + carry;

        result[i] = (sum % 2) + '0';
        carry = sum / 2;
    }

    if (carry)
        result[0] = '1';
    else {
        for (size_t i = 1; i < result_len; i++) {
            result[i-1] = result[i];
        }
    }

    return result;
}

File: my-folder/0067-add-binary/solution.rs
impl Solution {
    pub fn add_binary(a: String, b: String) -> String {
        let mut carry = 0;

        let mut a = a;
        let mut b = b;
        if a.len() < b.len() {
            std::mem::swap(&mut a, &mut b);
        }

        let result: String = a.chars()
            .rev()
            .zip(b.chars().rev().chain(std::iter::repeat('0')))
            .map(|(ac, bc)| {
                let sum = ac.to_digit(2).unwrap() + bc.to_digit(2).unwrap() + carry;
                carry = sum / 2;
                (sum % 2).to_string()
            })
            .collect();

        if carry != 0 {
            format!("{}{}", carry, result.chars().rev().collect::<String>())
        } else {
            result.chars().rev().collect()
        } 
    }
}
